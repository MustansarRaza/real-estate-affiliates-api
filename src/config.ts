import convict from "convict";
import dotenv from "dotenv";
import path from "path";

dotenv.config({
  path: path.join(process.cwd(), ".env"),
});

convict.addFormat({
  name: "OptionalString",
  validate: (value) => !value || typeof value === "string",
});

const config = convict({
  host: {
    doc: "Hostname/IP the service should bind on.",
    format: "String",
    default: "0.0.0.0",
    env: "HOST",
  },
  port: {
    doc: "Port number for the HTTP server to listen on.",
    format: "port",
    default: 5000,
    env: "PORT",
  },
  secret: {
    doc: "Secret to encrypt cookies with.",
    format: "*",
    default: "test",
    env: "APP_SECRET",
    sensitive: true,
  },
  enableCompression: {
    doc: "Whether to enable GZIP compression of response bodies or not.",
    format: "Boolean",
    default: true,
    env: "APP_ENABLE_COMPRESSION",
  },
  environment: {
    name: {
      doc: "Name of the environment we're running in.",
      format: "String",
      default: "unknown",
      env: "APP_ENVIRONMENT",
    },
    distribution: {
      doc: "Name/description of the build flavour we're running.",
      format: "String",
      default: "default",
      env: "APP_DIST",
    },
    release: {
      doc: "Name/description of the release version we're running.",
      format: "String",
      default: "unknown",
      env: "APP_RELEASE",
    },
  },
  keycloak: {
    realm: {
      doc: "Name of the keycloak realm to use.",
      format: "String",
      default: "propforce",
      env: "KEYCLOAK_REALM",
    },
    clientID: {
      doc: "ID of the client to authenticate with.",
      format: "String",
      default: "affiliates",
      env: "KEYCLOAK_CLIENT_ID",
    },
    clientSecret: {
      doc: "Secret key to authenticate as a client with.",
      format: "String",
      default: null,
      env: "KEYCLOAK_CLIENT_SECRET",
      sensitive: true,
    },
    authServerURL: {
      doc: "URL of the Keycloak server to use (note: ends in /auth).",
      format: "String",
      default: "https://stagemyaccount.propforce.com/auth",
      env: "KEYCLOAK_AUTH_SERVER_URL",
    },
  },
  db: {
    url: {
      doc: "MySQL database connection string",
      format: "String",
      default: "mysql://root:@127.0.0.1:3306/zameen_main",
      env: "DATABASE_URL",
      sensitive: true,
    },
  },
  cache: {
    url: {
      doc: "Redis connection string",
      format: "OptionalString",
      default: null,
      env: "REDIS_URL",
      sensitive: true,
    },
  },
  s3: {
    bucketName: {
      doc: "Name of the AWS S3 bucket to use for images.",
      format: "String",
      default: "affiliates-stage",
      env: "AWS_S3_BUCKET_NAME",
    },
    bucketRegion: {
      doc: "Name of the AWS region the S3 bucket is in.",
      format: "String",
      default: "eu-west-1",
      env: "AWS_S3_BUCKET_REGION",
    },
    accessKeyID: {
      doc: "AWS S3 public access key ID for images.",
      format: "String",
      default: null,
      env: "AWS_S3_ACCESS_KEY_ID",
    },
    secretAccessKey: {
      doc: "AWS S3 secret access key for images.",
      format: "String",
      default: null,
      env: "AWS_S3_SECRET_ACCESS_KEY",
      sensitive: true,
    },
  },
  sentry: {
    dsn: {
      doc: "Sentry's DSN to send events to.",
      format: "OptionalString",
      default: null,
      env: "SENTRY_DSN",
      sensitive: true,
    },
  },
  maxActiveAffiliateReservations: {
    doc: "Number of maximum active reservations for affiliates",
    format: "int",
    default: 2,
    env: "MAX_ACTIVE_AFFILIATE_RESERVATIONS",
  },
  dbTimezone: {
    doc: "Timezone of the database in hours",
    format: "String",
    default: "+05:00",
    env: "DB_TIMEZONE",
  },
});

const loadConfig = (): void => {
  config.validate({ allowed: "strict" });
};

export { config, loadConfig };
