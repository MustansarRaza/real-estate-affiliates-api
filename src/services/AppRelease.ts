import { AppRelease } from "affiliates-api/models";
import { Op } from "sequelize";

export interface AppVariablesDTO {
  min_version: string;
}
export interface SettingsDTO {
  version: string;
  appVariables: AppVariablesDTO;
}
class AppReleaseService {
  /**
   * Returns mobile application settings
   */
  async get(appName: string): Promise<SettingsDTO | null> {
    const release = await AppRelease.findOne({
      attributes: ["version", "appVariables"],
      where: {
        appName: {
          [Op.eq]: appName,
        },
      },
    });

    return release ? this.serialize(release) : null;
  }
  /**
   * Serializes a queried {@see AppRelease} model instance into a DTO.
   */
  private serialize(release: AppRelease): SettingsDTO {
    return {
      version: release.version.toString(),
      appVariables: {
        min_version: release.appVariables!.min_version,
      },
    };
  }
}
export default AppReleaseService;
