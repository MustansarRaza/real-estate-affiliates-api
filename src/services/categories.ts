import { ListingStatusKey, PlatformKey } from "@pf/taxonomies";
import {
  Address,
  AgencyConfiguration,
  Category,
  Listing,
  ListingStatus,
  Location,
  Platform,
  popularCategoryKeyToIndex,
  Project,
  ProjectAffiliateVisibilityID,
  ProjectStatus,
} from "affiliates-api/models";
import { col, fn, Op } from "sequelize";

interface CategoryAvailabilityDTO {
  availableInCities: CategoryCityAvailabilityDTO[];
  availableInProjects: number[];
}

interface CategoryCityAvailabilityDTO {
  cityID: number;
  unitsCount: number;
}

interface CategoryDTO {
  id: number;
  name: string;
  nameSingular: string;
  orderIndex: number;
  slug: string;
  parentID: number | null;
  popularityIndex: number;
  availableInCities: CategoryCityAvailabilityDTO[];
  availableInProjects: number[];
}

/**
 * Provides access to categories.
 */
class CategoriesService {
  /**
   * Gets a list of all available categories.
   */
  async list(agencyID: number | null): Promise<CategoryDTO[]> {
    let platformId = null;

    if (agencyID) {
      const agencyConfig = await AgencyConfiguration.findOne({
        attributes: ["platformId"],
        where: { agencyId: { [Op.eq]: agencyID } },
        raw: true,
      });
      platformId = agencyConfig?.platformId;
    }
    if (!platformId) {
      const platform = await Platform.findOne({
        attributes: ["id"],
        where: { platformKey: { [Op.eq]: PlatformKey.ZAMEEN } },
        raw: true,
      });
      platformId = platform.id;
    }

    const categories = await Category.findAll({
      where: { platformId: { [Op.eq]: platformId } },
      order: [
        ["type_order", "ASC"],
        ["type_id", "ASC"],
      ],
    });

    const categoryCitiesAndProjectsMap = await this.getCategoryCitiesAndProjectsMap(agencyID);

    return categories.map((category: Category) => ({
      id: category.id,
      name: category.name,
      nameSingular: category.nameSingular,
      orderIndex: category.orderIndex,
      slug: category.slug,
      parentID: category.parentID,
      popularityIndex: popularCategoryKeyToIndex(category.key),
      availableInCities: categoryCitiesAndProjectsMap[category.id]?.availableInCities || [],
      availableInProjects: categoryCitiesAndProjectsMap[category.id]?.availableInProjects || [],
    }));
  }

  /**
   * Gets a map of cities and Projects against a category in which
   * that category is available.
   */
  private async getCategoryCitiesAndProjectsMap(
    agencyID: number | null
  ): Promise<{
    [propertTypeId: number]: CategoryAvailabilityDTO;
  }> {
    const categoriesAvailablityResult = await Listing.findAll({
      attributes: [[fn("COUNT", "id"), "Count"]],
      include: [
        {
          model: ListingStatus,
          attributes: [],
          where: {
            key: {
              [Op.eq]: ListingStatusKey.AVAILABLE,
            },
          },
          required: true,
        },
        {
          model: Category,
          attributes: ["id"],
          required: true,
        },
        {
          model: Project,
          required: true,
          attributes: ["id"],
          where: {
            status: {
              [Op.eq]: ProjectStatus.ON,
            },
            affiliateVisibility: {
              [Op.eq]: ProjectAffiliateVisibilityID.VISIBLE,
            },
            ...(agencyID && {
              agencyID: {
                [Op.eq]: agencyID,
              },
            }),
          },
          include: [
            {
              model: Address,
              required: true,
              attributes: [],
              include: [
                {
                  model: Location,
                  attributes: ["id"],
                  required: true,
                },
              ],
            },
          ],
        },
      ],
      group: [
        col("Category.type_id"),
        col("Project.Address.Location.title"),
        col("Project.project_id"),
      ],
      raw: true,
    });

    const categoryCitiesAndProjectsMap: { [propertTypeId: number]: CategoryAvailabilityDTO } = {};
    categoriesAvailablityResult.forEach((row: any) => {
      const propertyTypeId = row["Category.id"];
      const projectID = row["Project.id"];
      const cityID = row["Project.Address.Location.id"];
      const unitsCount = row["Count"];

      if (categoryCitiesAndProjectsMap.hasOwnProperty(propertyTypeId)) {
        categoryCitiesAndProjectsMap[propertyTypeId].availableInProjects.push(projectID);
        const index = categoryCitiesAndProjectsMap[propertyTypeId].availableInCities
          .map((city: CategoryCityAvailabilityDTO) => city.cityID)
          .indexOf(cityID);
        if (index === -1) {
          categoryCitiesAndProjectsMap[propertyTypeId].availableInCities.push({
            cityID,
            unitsCount,
          });
        } else {
          categoryCitiesAndProjectsMap[propertyTypeId].availableInCities[
            index
          ].unitsCount += unitsCount;
        }
      } else {
        categoryCitiesAndProjectsMap[propertyTypeId] = {
          availableInProjects: [projectID],
          availableInCities: [{ cityID, unitsCount }],
        };
      }
    });

    return categoryCitiesAndProjectsMap;
  }
}

export default CategoriesService;
