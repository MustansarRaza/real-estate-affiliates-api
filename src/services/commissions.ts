import { AffiliateCommissionType } from "@pf/taxonomies";
import { AffiliateCommission } from "affiliates-api/models";
import { maxBy, minBy } from "lodash";
import { Op, WhereOptions } from "sequelize";

export interface ProjectCommissionDTO {
  max: number | null;
  min: number | null;
}

export interface AffiliateCommissionDTO {
  [commissionType: string]: {
    [propertyTypeId: number]: AffiliateCommission[];
  };
}

export interface AffiliateProjectCommissionDTO {
  [projectID: number]: AffiliateCommissionDTO;
}

type ByProjectIDLookup = {
  projectIDs: number[];
};

type ByAffiliateTypeIDLookup = {
  affiliateTypeId: number | null;
};

type ByFiltersLookup = {
  createdBy?: number;
  propertyTypeId?: number;
  commissionType?: AffiliateCommissionType;
};

type AffiliateCommissionLookupOptions = ByProjectIDLookup &
  ByAffiliateTypeIDLookup &
  ByFiltersLookup;

class CommissionsService {
  async getDetailsFor(
    options: AffiliateCommissionLookupOptions
  ): Promise<{ [projectID: number]: AffiliateCommission[] }> {
    const unformattedDate = new Date();
    unformattedDate.setHours(0, 0, 0, 0);
    const currentDate = unformattedDate.toISOString();

    const commissionDetails = await AffiliateCommission.findAll({
      where: {
        [Op.and]: this.createFiltersForAffiliateCommission(options),
        [Op.or]: [
          {
            startDate: {
              [Op.is]: null,
            },
            endDate: {
              [Op.is]: null,
            },
          },
          {
            startDate: {
              [Op.not]: null,
              [Op.lte]: currentDate,
            },
            endDate: {
              [Op.is]: null,
            },
          },
          {
            startDate: {
              [Op.is]: null,
            },
            endDate: {
              [Op.not]: null,
              [Op.gte]: currentDate,
            },
          },
          {
            startDate: {
              [Op.not]: null,
              [Op.lte]: currentDate,
            },
            endDate: {
              [Op.not]: null,
              [Op.gte]: currentDate,
            },
          },
        ],
      },
      raw: true,
    });

    const projectAndCommissionsMapping: { [projectID: number]: AffiliateCommission[] } = {};
    commissionDetails.forEach((row: AffiliateCommission) => {
      if (projectAndCommissionsMapping.hasOwnProperty(row.projectId))
        projectAndCommissionsMapping[row.projectId].push(row);
      else projectAndCommissionsMapping[row.projectId] = [row];
    });

    return projectAndCommissionsMapping;
  }

  async getCommissionsBreakdownForAffiliate(
    options: AffiliateCommissionLookupOptions
  ): Promise<AffiliateProjectCommissionDTO> {
    const projectAndCommissionsMapping = await this.getDetailsFor(options);

    const avaiableCommissionProjectIDs = Object.keys(projectAndCommissionsMapping);

    const affiliateCommissionBreakdown: AffiliateProjectCommissionDTO = {};
    avaiableCommissionProjectIDs.forEach((projectID) => {
      const projectCommissions = projectAndCommissionsMapping[parseInt(projectID, 10)] || [];

      projectCommissions.forEach((commission) => {
        if (affiliateCommissionBreakdown.hasOwnProperty(commission.projectId))
          if (
            affiliateCommissionBreakdown[commission.projectId].hasOwnProperty(
              commission.commissionType
            )
          )
            if (
              affiliateCommissionBreakdown[commission.projectId][
                commission.commissionType
              ].hasOwnProperty(commission.propertyTypeId)
            )
              affiliateCommissionBreakdown[commission.projectId][commission.commissionType][
                commission.propertyTypeId
              ].push(commission);
            else
              affiliateCommissionBreakdown[commission.projectId][commission.commissionType][
                commission.propertyTypeId
              ] = [commission];
          else
            affiliateCommissionBreakdown[commission.projectId][commission.commissionType] = {
              [commission.propertyTypeId]: [commission],
            };
        else
          affiliateCommissionBreakdown[commission.projectId] = {
            [commission.commissionType]: { [commission.propertyTypeId]: [commission] },
          };
      });
    });

    return affiliateCommissionBreakdown;
  }

  async getProjectCommissionStats(
    options: AffiliateCommissionLookupOptions
  ): Promise<Map<number, ProjectCommissionDTO>> {
    const projectAndCommissionsMapping = await this.getDetailsFor(options);

    const projectCommissionStats = new Map<number, ProjectCommissionDTO>();
    Object.keys(projectAndCommissionsMapping).forEach((projectID) => {
      const maxCommission = maxBy(
        projectAndCommissionsMapping[+projectID],
        (affiliateCommission) =>
          affiliateCommission.commission + affiliateCommission.classifiedCredits!
      );
      const minCommission = minBy(
        projectAndCommissionsMapping[+projectID],
        (affiliateCommission) =>
          affiliateCommission.commission + affiliateCommission.classifiedCredits!
      );

      return projectCommissionStats.set(+projectID, {
        max: maxCommission?.commission! + maxCommission?.classifiedCredits!,
        min: minCommission?.commission! + minCommission?.classifiedCredits!,
      });
    });

    return projectCommissionStats;
  }

  private createFiltersForAffiliateCommission(
    options: AffiliateCommissionLookupOptions
  ): WhereOptions[] {
    const filters: WhereOptions[] = [
      {
        affiliateTypeId: {
          [Op.eq]: options.affiliateTypeId,
        },
        projectId: {
          [Op.in]: options.projectIDs,
        },
      },
    ];

    if (options?.createdBy) {
      filters.push({
        createdBy: {
          [Op.eq]: options.createdBy,
        },
      });
    }
    if (options?.propertyTypeId) {
      filters.push({
        propertyTypeId: {
          [Op.eq]: options.propertyTypeId,
        },
      });
    }
    if (options?.commissionType) {
      filters.push({
        commissionType: {
          [Op.eq]: options.commissionType,
        },
      });
    }

    return filters;
  }
}

export default CommissionsService;
