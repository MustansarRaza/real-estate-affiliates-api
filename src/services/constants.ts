export enum ErrorMessages {
  USER_AFFILIATE_ERROR_MESSAGE = "Current user is not an affiliate",
}

export const FORMAT_DATE = {
  FULL_DATE_24_HOUR_FORMAT_DATE_FNS: "YYYY-MM-DD HH:mm:ss",
};
