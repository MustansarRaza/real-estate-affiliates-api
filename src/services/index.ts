import AppRelease from "./AppRelease";
import CategoriesService from "./categories";
import ClientsService, { CreateClientDTO as CreateClientDTOType } from "./clients";
import CommissionsService from "./commissions";
import LeadsService, { CreateLeadDTO as CreateLeadDTOType } from "./leads";
import ListingsService, { ListingDTO as ListingDTOType } from "./listings";
import LocationsService from "./locations";
import MeService, { CreateMeDTO as CreateMeDTOType, MeDTO as MeDTOType } from "./me";
import ProfileService from "./profile";
import ProjectsService from "./projects";
import ListingReservationService, { ListingReservationOptions } from "./reservations";
import TasksService from "./tasks";
import TaskTypesService from "./taskTypes";

export interface ServiceMesh {
  profile: ProfileService;
  me: MeService;
  clients: ClientsService;
  tasks: TasksService;
  taskTypes: TaskTypesService;
  leads: LeadsService;
  projects: ProjectsService;
  listings: ListingsService;
  locations: LocationsService;
  categories: CategoriesService;
  reservations: ListingReservationService;
  commissions: CommissionsService;
  appRelease: AppRelease;
}

const useServices = (): ServiceMesh => {
  const locations = new LocationsService();
  const profile = new ProfileService({ locations });
  const me = new MeService({ profile, locations });
  const clients = new ClientsService({ me, locations });
  const taskTypes = new TaskTypesService();
  const leads = new LeadsService({ me, taskTypes });
  const tasks = new TasksService({ leads, taskTypes });
  const listings = new ListingsService({ me });
  const commissions = new CommissionsService();
  const projects = new ProjectsService({ me, listings, commissions, locations });
  const categories = new CategoriesService();
  const reservations = new ListingReservationService({ me, listings });
  const appRelease = new AppRelease();

  return {
    profile,
    me,
    clients,
    leads,
    tasks,
    taskTypes,
    projects,
    listings,
    locations,
    categories,
    reservations,
    commissions,
    appRelease,
  };
};

export {
  useServices,
  ClientsService,
  LeadsService,
  MeService,
  ProfileService,
  ProjectsService,
  ListingsService,
  TasksService,
  TaskTypesService,
  LocationsService,
  CategoriesService,
  ListingReservationService,
  CommissionsService,
  AppRelease,
};

export type CreateClientDTO = CreateClientDTOType;
export type CreateLeadDTO = CreateLeadDTOType;
export type CreateMeDTO = CreateMeDTOType;
export type CreateReservationDTO = ListingReservationOptions;
export type ListingDTO = ListingDTOType;
export type MeDTO = MeDTOType;
