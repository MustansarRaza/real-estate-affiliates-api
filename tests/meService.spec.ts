import { AppError } from "@pf/utils";
import { config } from "affiliates-api/config";
import {
  Affiliate,
  AffiliateAssignee,
  AffiliateAssigneeType,
  AffiliateType,
  GovIDType,
} from "affiliates-api/models";
import { MeService } from "affiliates-api/services";

import { Op } from "sequelize";
import { loadSequelizeFixtures, useKeycloakMock, useMockedServices } from "./support";
import { ModelFixtureConstants, ModelFixtureNames } from "./support/data";

describe("MeService", () => {
  const mockedServices = useMockedServices();
  const service = new MeService(mockedServices);

  describe("create", () => {
    const mockLocation = {
      id: 3,
      parentID: null,
      name: "Lahore",
      level1Location: {
        id: 1,
        name: "Pakistan",
        parentID: null,
      },
    };

    const createMeBaseDTO = {
      profile: {
        email: "walt@disney.com",
        givenName: "Walt",
        familyName: "Disney",
        mobilePhoneNumbers: ["+923377178218"],
        landLinePhoneNumbers: ["+922123569856"],
        address: "Walstreet 23",
        govIDType: GovIDType.NIC,
        govIDNumber: "2382738923",
        taxNumber: "2349231292",
        businessName: "Disney",
        affiliateType: AffiliateType.AGENCY,
        locationID: mockLocation.id,
        termsAndConditionsVersion: "1.1.1",
      },
      credentials: {
        password: "test",
      },
    };

    beforeEach(() => {
      jest.resetAllMocks();

      mockedServices.locations.getWithHierarchy.mockReturnValue(Promise.resolve(mockLocation));
    });

    it("creates a new user", async () => {
      mockedServices.profile.create.mockReturnValueOnce(
        Promise.resolve({ username: "walt@disney.com", ...createMeBaseDTO.profile })
      );

      const createdUser = await service.create(createMeBaseDTO);

      // replacing ID because it changes on every run
      expect(createdUser.ssoID).toBeTruthy();
      expect({ ...createdUser, ssoID: "test" }).toMatchSnapshot();

      // make sure the user was created in keycloak
      const kmock = useKeycloakMock();
      const keycloakUser = kmock.database.findUserByID(createdUser.ssoID);
      expect(keycloakUser?.profile.id).toBe(createdUser.ssoID);
      expect(keycloakUser?.profile.username).toBe(createMeBaseDTO.profile.email);
      expect(keycloakUser?.profile.email).toBe(createMeBaseDTO.profile.email);
      expect(keycloakUser?.profile.enabled).toBe(true);
      expect(keycloakUser?.profile.emailVerified).toBe(true);
      expect(keycloakUser?.profile.firstName).toBe(createMeBaseDTO.profile.givenName);
      expect(keycloakUser?.profile.lastName).toBe(createMeBaseDTO.profile.familyName);
      expect(keycloakUser?.profile.attributes).toMatchSnapshot();
      expect(keycloakUser?.credentials).toHaveLength(1);
      expect(keycloakUser?.credentials[0].value).toBe("test");
    });

    it("throws an error when creating the same user twice", async () => {
      mockedServices.profile.find.mockReturnValue(Promise.resolve(null));

      mockedServices.profile.create.mockReturnValue(
        Promise.resolve({ ...createMeBaseDTO.profile, username: "walt@disney.com" })
      );

      const createdUser = await service.create(createMeBaseDTO);

      mockedServices.profile.find.mockReturnValue(
        Promise.resolve({ ...createMeBaseDTO.profile, username: "walt@disney.com" })
      );

      await expect(service.create(createMeBaseDTO)).rejects.toMatchSnapshot();

      // make sure that the user in keycloak didn't get re-created
      const kmock = useKeycloakMock();
      const keycloakUsers = kmock.database.allUsers();
      expect(keycloakUsers).toHaveLength(1);
      expect(keycloakUsers[0]?.profile.id).toBe(createdUser.ssoID);
    });

    it("allows re-creating the user when a previous attempt failed to create profile", async () => {
      const kmock = useKeycloakMock();

      mockedServices.profile.find.mockReturnValue(Promise.resolve(null));
      mockedServices.profile.create.mockImplementation(() => {
        throw new Error("crap!");
      });

      // user creation will fail because profile creation failed,
      // but we'll end up with a dead user in keycloak
      await expect(service.create(createMeBaseDTO)).rejects.toMatchSnapshot();

      const keycloakUsersAfterFailure = kmock.database.allUsers();
      expect(keycloakUsersAfterFailure).toHaveLength(1);

      // now, we want to do another attempt, it should work
      mockedServices.profile.create.mockReturnValue(
        Promise.resolve({ username: "walt@disney.com", ...createMeBaseDTO.profile })
      );

      const createdUser = await service.create(createMeBaseDTO);
      expect(createdUser).toBeTruthy();

      const keycloakUsersAfterSuccess = kmock.database.allUsers();
      expect(keycloakUsersAfterSuccess).toHaveLength(1);

      // make sure the keycloak user got re-created
      const keycloakUserAfterFailure = keycloakUsersAfterFailure[0];
      const keycloakUserAfterSuccess = keycloakUsersAfterSuccess[0];
      expect(keycloakUserAfterFailure.profile.id).not.toBe(keycloakUserAfterSuccess.profile.id);

      expect(createdUser.ssoID).toBe(keycloakUserAfterSuccess.profile.id);
    });

    it("lets errors during profile creation bubble up", async () => {
      const error = new AppError("brrr stuff goes brrr", {
        statusCode: 400,
      });

      mockedServices.profile.create.mockImplementation(() => {
        throw error;
      });

      await expect(service.create(createMeBaseDTO)).rejects.toThrow(error);
    });

    it("raises an error if keycloak client secret is missing", async () => {
      config.set("keycloak.clientSecret", null);

      await expect(service.create(createMeBaseDTO)).rejects.toMatchSnapshot();
    });

    it("raises an error if an invalid location ID is specified", async () => {
      mockedServices.locations.getWithHierarchy.mockReturnValue(Promise.resolve(null));

      await expect(service.create(createMeBaseDTO)).rejects.toMatchSnapshot();
    });

    it("raises an error if a location without a level1 ID is specified", async () => {
      mockedServices.locations.getWithHierarchy.mockReturnValue(
        Promise.resolve({
          ...mockLocation,
          level1Location: null,
        })
      );

      await expect(service.create(createMeBaseDTO)).rejects.toMatchSnapshot();
    });
  });

  describe("assignees", () => {
    it("returns a list of assignees for the specified SSO ID", async () => {
      await loadSequelizeFixtures(
        ModelFixtureNames.AGENTS,
        ModelFixtureNames.CONTACTS,
        ModelFixtureNames.AFFILIATES,
        ModelFixtureNames.AFFILIATE_CONTACTS,
        ModelFixtureNames.AFFILIATE_ASSIGNEES,
        ModelFixtureNames.POTENTIAL_AFFILIATE_VERIFICATION_STATUS,
        ModelFixtureNames.POTENTIAL_AFFILIATES,
        ModelFixtureNames.USERS
      );

      const assignees = await service.assignees({ ssoID: ModelFixtureConstants.verifiedSSOID });
      expect(assignees).toMatchSnapshot();
    });

    it("returns an empty array if no user exists with the specified sso ID", async () => {
      const ssoID = "bc9ddae1-4db9-41ed-b511-e3f08e5ba3e1";
      const assignees = await service.assignees({ ssoID });
      expect(assignees).toHaveLength(0);
    });
  });

  describe("owner", () => {
    it("returns the assignee that is the BDM for the specified SSO user", async () => {
      await loadSequelizeFixtures(
        ModelFixtureNames.AGENTS,
        ModelFixtureNames.USERS,
        ModelFixtureNames.CONTACTS,
        ModelFixtureNames.AFFILIATES,
        ModelFixtureNames.AFFILIATE_CONTACTS,
        ModelFixtureNames.AFFILIATE_ASSIGNEES
      );

      const owner = await service.owner({ ssoID: ModelFixtureConstants.verifiedSSOID });
      expect(owner?.userID).toBe(ModelFixtureConstants.staffUserBDMID);
    });

    it("returns the agency ID from the affiliate", async () => {
      await loadSequelizeFixtures(
        ModelFixtureNames.AGENTS,
        ModelFixtureNames.USERS,
        ModelFixtureNames.CONTACTS,
        ModelFixtureNames.AFFILIATES,
        ModelFixtureNames.AFFILIATE_CONTACTS,
        ModelFixtureNames.AFFILIATE_ASSIGNEES
      );

      const affiliate = await Affiliate.findOne({
        where: {
          id: {
            [Op.eq]: ModelFixtureConstants.verifiedAffiliateID,
          },
        },
      });

      const owner = await service.owner({ ssoID: ModelFixtureConstants.verifiedSSOID });
      expect(owner?.agencyID).toBe(affiliate.agencyID);
    });

    it("returns the first BDM assignee, ordered by primary key", async () => {
      await loadSequelizeFixtures(
        ModelFixtureNames.AGENTS,
        ModelFixtureNames.USERS,
        ModelFixtureNames.CONTACTS,
        ModelFixtureNames.AFFILIATES,
        ModelFixtureNames.AFFILIATE_CONTACTS
      );

      await AffiliateAssignee.create({
        id: 2,
        affiliateID: ModelFixtureConstants.verifiedAffiliateID,
        type: AffiliateAssigneeType.BDM,
        userID: ModelFixtureConstants.staffUserBDMID,
      });

      await AffiliateAssignee.create({
        id: 1,
        affiliateID: ModelFixtureConstants.verifiedAffiliateID,
        type: AffiliateAssigneeType.BDM,
        userID: ModelFixtureConstants.staffUserCoordinatorID,
      });

      const owner = await service.owner({ ssoID: ModelFixtureConstants.verifiedSSOID });
      expect(owner?.userID).toBe(ModelFixtureConstants.staffUserCoordinatorID);
    });

    it("returns null if no BDM assignee could be found", async () => {
      await loadSequelizeFixtures(
        ModelFixtureNames.AGENTS,
        ModelFixtureNames.USERS,
        ModelFixtureNames.CONTACTS,
        ModelFixtureNames.AFFILIATES,
        ModelFixtureNames.AFFILIATE_CONTACTS
      );

      const owner = await service.owner({ ssoID: ModelFixtureConstants.verifiedSSOID });
      expect(owner).toBe(null);
    });
  });

  describe("acceptTermsAndConditions", () => {
    const profileDTO = {
      profile: {
        email: "walt@disney.com",
        givenName: "Walt",
        familyName: "Disney",
        mobilePhoneNumbers: ["+923377178218"],
        landLinePhoneNumbers: ["+922123569856"],
        address: "Walstreet 23",
        govIDType: GovIDType.NIC,
        govIDNumber: "2382738923",
        taxNumber: "2349231292",
        businessName: "Disney",
        affiliateType: AffiliateType.AGENCY,
        locationID: 3,
        termsAndConditionsVersion: "1.1.1",
      },
      credentials: {
        password: "test",
      },
    };
    beforeEach(async () => {
      jest.resetAllMocks();

      await loadSequelizeFixtures(
        ModelFixtureNames.AGENTS,
        ModelFixtureNames.USERS,
        ModelFixtureNames.CONTACTS,
        ModelFixtureNames.AFFILIATES,
        ModelFixtureNames.AFFILIATE_CONTACTS,
        ModelFixtureNames.POTENTIAL_AFFILIATE_VERIFICATION_STATUS,
        ModelFixtureNames.POTENTIAL_AFFILIATES,
        ModelFixtureNames.AFFILIATE_ASSIGNEES
      );

      mockedServices.profile.get.mockReturnValue(
        Promise.resolve({ ...profileDTO.profile, username: "walt@disney.com" })
      );
    });
    it("accepts the Terms and conditions for an already created user", async () => {
      const updatedMe = await service.acceptTermsAndConditions(
        ModelFixtureConstants.verifiedSSOID,
        "1.1.1"
      );

      expect(updatedMe.profile.termsAndConditionsVersion).not.toBeNull();
      expect(updatedMe.profile.termsAndConditionsVersion).toEqual("1.1.1");
    });

    it("raises an error if ssoID not found", async () => {
      const error = new AppError("Current user is not an affiliate", {
        statusCode: 404,
      });

      await expect(
        service.acceptTermsAndConditions(ModelFixtureConstants.nonExistantSSOID, "1.1.1")
      ).rejects.toThrow(error);
    });
  });
});
