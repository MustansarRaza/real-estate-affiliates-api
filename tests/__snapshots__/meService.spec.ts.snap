// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeService assignees returns a list of assignees for the specified SSO ID 1`] = `
Array [
  Object {
    "affiliateID": 1,
    "type": "bdm",
    "user": Object {
      "email": "henk@world.com",
      "hasWhatsApp": true,
      "id": 1,
      "landLinePhoneNumber": "+922123569856",
      "mobilePhoneNumber": "+923377178218",
      "name": "Henk Jansen",
    },
  },
  Object {
    "affiliateID": 1,
    "type": "coordinator",
    "user": Object {
      "email": "walt@disney.com",
      "hasWhatsApp": false,
      "id": 2,
      "landLinePhoneNumber": null,
      "mobilePhoneNumber": "+923377178218",
      "name": "Walt Disney",
    },
  },
]
`;

exports[`MeService create allows re-creating the user when a previous attempt failed to create profile 1`] = `[Error: crap!]`;

exports[`MeService create creates a new user 1`] = `
Object {
  "affiliate": null,
  "affiliateVerificationStatus": Object {
    "id": 1,
    "status": "new",
  },
  "profile": Object {
    "address": "Walstreet 23",
    "affiliateType": "agency",
    "businessName": "Disney",
    "email": "walt@disney.com",
    "familyName": "Disney",
    "givenName": "Walt",
    "govIDNumber": "2382738923",
    "govIDType": "NIC",
    "landLinePhoneNumbers": Array [
      "+922123569856",
    ],
    "locationID": 3,
    "mobilePhoneNumbers": Array [
      "+923377178218",
    ],
    "taxNumber": "2349231292",
    "termsAndConditionsVersion": "1.1.1",
    "username": "walt@disney.com",
  },
  "ssoID": "test",
  "status": "unverified",
}
`;

exports[`MeService create creates a new user 2`] = `
Object {
  "address": Array [
    "Walstreet 23",
  ],
  "affiliateType": Array [
    1,
  ],
  "affiliateTypeTitle": Array [
    "Agency",
  ],
  "businessAddress": Array [
    "Walstreet 23",
  ],
  "businessName": Array [
    "Disney",
  ],
  "city": Array [
    3,
  ],
  "country": Array [
    1,
  ],
  "govtId": Array [
    "NIC",
  ],
  "isProfileCompleted": Array [
    "yes",
  ],
  "landline": Array [
    "+922123569856",
  ],
  "mobile": Array [
    "+923377178218",
  ],
  "ntn": "2349231292",
}
`;

exports[`MeService create raises an error if a location without a level1 ID is specified 1`] = `
Object {
  "extra": Object {
    "locationID": 3,
  },
  "message": "Cannot find location",
  "name": "BadRequest",
  "originalError": undefined,
  "statusCode": 400,
}
`;

exports[`MeService create raises an error if an invalid location ID is specified 1`] = `
Object {
  "extra": Object {
    "locationID": 3,
  },
  "message": "Cannot find location",
  "name": "BadRequest",
  "originalError": undefined,
  "statusCode": 400,
}
`;

exports[`MeService create raises an error if keycloak client secret is missing 1`] = `
Object {
  "extra": Object {
    "clientID": "affiliates",
  },
  "message": "Keycloak client secret is not configured",
  "name": "WrongSetup",
  "originalError": undefined,
  "statusCode": 500,
}
`;

exports[`MeService create throws an error when creating the same user twice 1`] = `
Object {
  "extra": Object {
    "email": "walt@disney.com",
  },
  "message": "User with this e-mail or phone number already exists",
  "name": "Conflict",
  "originalError": undefined,
  "statusCode": 409,
}
`;
